console/v1/types_console_plugin.go

```diff
@@ -0,0 +1,204 @@
+package v1
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +openshift:compatibility-gen:level=1
+
+// ConsolePlugin is an extension for customizing OpenShift web console by
+// dynamically loading code from another service running on the cluster.
+//
+// Compatibility level 1: Stable within a major release for a minimum of 12 months or 3 minor releases (whichever is longer).
+type ConsolePlugin struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+
+	// +kubebuilder:validation:Required
+	Spec ConsolePluginSpec `json:"spec"`
+}
+
+// ConsolePluginSpec is the desired plugin configuration.
+type ConsolePluginSpec struct {
+	// displayName is the display name of the plugin.
+	// The dispalyName should be between 1 and 128 characters.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=128
+	DisplayName string `json:"displayName"`
+	// service is a Kubernetes Service that exposes the plugin using a
+	// deployment with an HTTP server. The Service must use HTTPS and
+	// Service serving certificate. The console backend will proxy the
+	// plugins assets from the Service using the service CA bundle.
+	// +kubebuilder:validation:Required
+	Service ConsolePluginService `json:"service"`
+	// proxy is a list of proxies that describe various service type
+	// to which the plugin needs to connect to.
+	// +optional
+	Proxy []ConsolePluginProxy `json:"proxy,omitempty"`
+	// i18n is the configuration of plugin's localization resources.
+	// +kubebuilder:validation:Required
```

Comment:
@spadgett if we make this optional, then I think we should drop the `Lazy` `LoadType`, and just keep the `Preload`, since making it optional would leave the loading to lazy-loading, if the value was not set to `Preload`

Reply:
You should still keep the option else users won't be able to explicitly choose `Lazy`. What would happen if you changed the default later?

Reply:
@JoelSpeed right, so lets keep the `Lazy` type, leave the field required, so user has to explicitly set the value. Thoughts ?

Reply:
I would make the field optional, you shouldn't need to default it, you can have optional fields and optional structs and then if the field isn't set (is `""`) the controller can default the behaviour. The benefit here is that then you can change the default later should you need to